common/db_test.go

```diff
@@ -578,3 +582,311 @@ func TestDBUnbondingLocks(t *testing.T) {
+
+	assert := assert.New(t)
+	assert.Equal(ticket.Sender.Hex(), actualSender)
+	assert.Equal(ticket.Recipient.Hex(), actualRecipient)
+	assert.Equal(ticket.FaceValue, new(big.Int).SetBytes(actualFaceValueBytes))
+	assert.Equal(ticket.WinProb, new(big.Int).SetBytes(actualWinProbBytes))
+	assert.Equal(ticket.SenderNonce, actualSenderNonce)
+	assert.Equal(recipientRand, new(big.Int).SetBytes(actualRecipientRandBytes))
+	assert.Equal(ticket.RecipientRandHash, ethcommon.HexToHash(actualRecipientRandHash))
+	assert.Equal(sig, actualSig)
+	assert.Equal(sessionID, actualSessionID)
+
+	ticketsCount := getRowCountOrFatal("SELECT count(*) FROM winningTickets", dbraw, t)
+	assert.Equal(1, ticketsCount)
+}
+
+func TestInsertWinningTicket_GivenMaxValueInputs_InsertsOneRowCorrectly(t *testing.T) {
+	dbh, dbraw, err := TempDB(t)
+	defer dbh.Close()
+	defer dbraw.Close()
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	sessionID, ticket, sig, recipientRand := defaultWinningTicket(t)
+	ticket.FaceValue = maxUint256OrFatal(t)
+	ticket.WinProb = maxUint256OrFatal(t)
+	ticket.SenderNonce = math.MaxUint32
+
+	err = dbh.StoreWinningTicket(sessionID, ticket, sig, recipientRand)
+
+	if err != nil {
+		t.Errorf("Failed to insert winning ticket with error: %v", err)
+	}
+
+	row := dbraw.QueryRow("SELECT sender, recipient, faceValue, winProb, senderNonce, recipientRand, recipientRandHash, sig, sessionID FROM winningTickets")
+	var actualSender, actualRecipient, actualRecipientRandHash, actualSessionID string
+	var actualFaceValueBytes, actualWinProbBytes, actualRecipientRandBytes, actualSig []byte
+	var actualSenderNonce uint32
+	err = row.Scan(&actualSender, &actualRecipient, &actualFaceValueBytes, &actualWinProbBytes, &actualSenderNonce, &actualRecipientRandBytes, &actualRecipientRandHash, &actualSig, &actualSessionID)
+
+	assert := assert.New(t)
+	assert.Equal(ticket.Sender.Hex(), actualSender)
+	assert.Equal(ticket.Recipient.Hex(), actualRecipient)
+	assert.Equal(ticket.FaceValue, new(big.Int).SetBytes(actualFaceValueBytes))
+	assert.Equal(ticket.WinProb, new(big.Int).SetBytes(actualWinProbBytes))
+	assert.Equal(ticket.SenderNonce, actualSenderNonce)
+	assert.Equal(recipientRand, new(big.Int).SetBytes(actualRecipientRandBytes))
+	assert.Equal(ticket.RecipientRandHash, ethcommon.HexToHash(actualRecipientRandHash))
+	assert.Equal(sig, actualSig)
+	assert.Equal(sessionID, actualSessionID)
+
+	ticketsCount := getRowCountOrFatal("SELECT count(*) FROM winningTickets", dbraw, t)
+	assert.Equal(1, ticketsCount)
+}
+
+func TestStoreWinningTicket_GivenNilTicket_ReturnsError(t *testing.T) {
+	dbh, dbraw, err := TempDB(t)
+	defer dbh.Close()
+	defer dbraw.Close()
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	sig := randBytesOrFatal(42, t)
+	recipientRand := new(big.Int).SetInt64(1234)
+
+	err = dbh.StoreWinningTicket("sessionID", nil, sig, recipientRand)
+
+	assert := assert.New(t)
+	assert.NotNil(err)
```

Comment:
IIUC `assert` will not stop test execution so in cases like this if this line fails, we would continue executing the next line which would always fail since `err = nil`. Should we try to avoid the additional assertion errors by structuring the assertions like this:

```
if assert.NotNil(err) {
    assert.Contains(err.Error(), "nil ticket")
}
```

Reply:
My approach so far is:

- Use `require` or `Fatal` for all errors that are part of the code the test function is relying on, i.e. assumes it works perfectly.
- Use `assert` for any errors that arise from the "code under test", since in this case it's also interesting to learn if we fail due to other errors.

Does this line of thinking make sense to you, or do you think we should think of this differently?

Reply:
That makes sense! Using `require` seems cleaner than nested asserts